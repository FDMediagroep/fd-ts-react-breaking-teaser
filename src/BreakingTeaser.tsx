import React, { PureComponent } from "react";
import { createGlobalStyle } from "styled-components";
import Card, { getAllCardStyles, CardTypes } from "@fdmg/fd-card";
import TypoGraphy, { getAllTextStyles } from "@fdmg/fd-typography";
import {
    RelatedArticle,
    SourceSet,
    TeaserFigure,
    TeaserFigureStyle,
    TeaserRelated,
    TeaserRelatedStyle,
    ImageType} from "@fdmg/fd-teaser";

export interface Props {
    bookmarked?: boolean;
    bookmarkTitle?: string;
    cardStyle?: CardTypes;
    className?: string;
    comments?: number;
    cropRectangle?: string;
    description?: string;
    hashTags?: string;
    onBookmark?: (e: React.MouseEvent<HTMLElement>) => void;
    related?: RelatedArticle[];
    id: string;
    image: ImageType;
    sourceSets: SourceSet[];
    readableAge: string;
    removeBookmarkTitle?: string;
    shareTitle?: string;
    subject: string;
    title: string;
    url: string;
}

export default class BreakingTeaser extends PureComponent<Props, any> {
    render() {
        return (
            <>
                <GlobalStyle/>
                <Card cardStyle={this.props.cardStyle ? this.props.cardStyle : 'default'} className={`fd-breaking-teaser${this.props.className ? ` ${this.props.className}` : ''}`}>
                    <div className="top-bar">
                        <span>Breaking</span>
                        <span className="icons">
                            <a className="share-social" href="javascript: void(0);">
                                <i
                                    data-url={this.props.url}
                                    data-id={this.props.id}
                                    data-title={this.props.title}
                                    data-section={this.props.subject}
                                    data-hashtags={this.props.hashTags}
                                    className="share-social icon-share-alt"
                                    title={this.props.shareTitle ? this.props.shareTitle : 'Delen'}
                                />
                            </a>
                            <a
                                href="javascript: void(0);"
                                className={`favorite${this.props.bookmarked ? ' selected' : ''}`}
                                data-key={this.props.id}
                                data-addurl='/add-favorite'
                                data-deleteurl='/delete-favorite'
                                data-title={this.props.title}
                                data-url={this.props.url}
                                data-articleid={this.props.id}
                                title="Bewaren"
                                id={`favorite${this.props.id}`}
                                onClick={this.props.onBookmark}
                            >
                                <i className="icon-bookmark" title={this.props.bookmarkTitle ? this.props.bookmarkTitle : 'Bewaren'}/>
                                <i className="icon-bookmark1" title={this.props.removeBookmarkTitle ? this.props.removeBookmarkTitle : 'Verwijder van bewaarlijst'}/>
                            </a>
                            <a className="comments-link" href={`${this.props.url}#comments`}>
                                {this.props.comments ? (
                                    <>
                                        <i className="icon-comment-alt"/>
                                        <span className="comments-count">{this.props.comments}</span>
                                        <span className="comments-suffix">{this.props.comments > 1 ? 'reacties' : 'reactie'}</span>
                                    </>
                                ) : null}
                            </a>
                        </span>
                    </div>
                    <a className="teaser-link" href={this.props.url}>
                        <TeaserFigure
                            image={this.props.image}
                            sourceSets={this.props.sourceSets}
                        />

                        <div className="text-container">
                            <TypoGraphy textStyle="breaking-teaser-h"><h2>{this.props.title}</h2></TypoGraphy>
                            {this.props.description ? <p className="intro">{this.props.description}</p> : null}
                        </div>

                    </a>
                    <TeaserRelated items={this.props.related}/>
                </Card>
            </>
        );
    }
}

const GlobalStyle = createGlobalStyle`
.fd-breaking-teaser {
    font-family: 'ProximaNovaRegular', Helvetica, sans-serif;

    .top-bar,
    .breaking-teaser-h,
    .intro {
        padding: 0 10px;
        @media only screen and (min-width: 641px) {
            padding: 0 12px;
        }
        @media only screen and (min-width: 861px) {
            padding: 0 15px;
        }
    }

    a {
        text-decoration: none;
        color: #191919;
    }

    .top-bar {
        display: flex;
        justify-content: space-between;
        align-items: center;
        background-color: #c60f0f;
        color: #ffffff;
        font-family: 'ProximaNovaBold', sans-serif;
        min-height: 44px;
        line-height: 1.1em;
        font-size: 1rem;
        @media only screen and (min-width: 641px) {
            font-size: 1.125rem;
        }

        i {
            cursor: pointer;
            padding: .5rem;
        }

        .share-social {
            color: #ffffff;
        }

        .favorite {
            color: #ffffff;
            padding: 0;
            .icon-bookmark1 {
                display: none;
            }
            &.selected {
                .icon-bookmark {
                    display: none;
                }
                .icon-bookmark1 {
                    display: initial;
                }
            }
        }

        .comments-link {
            display: inline-flex;
            align-items: center;
            color: #ffffff;
            font-family: 'ProximaNovaRegular', sans-serif;
            i {
                padding-right: 0;
            }
            span {
                font-size: 0.8125rem;
                cursor: pointer;
            }
            .comments-count,
            .comments-suffix {
                margin-left: .2rem;
            }
            .comments-suffix {
                @media only screen and (max-width: 640px) {
                    display: none;
                }
            }
        }
    }

    figure {
        margin: 0;
        img {
            width: 100%;
        }
        line-height: 0;
    }

    .teaser-link {
        &:hover {
            .breaking-teaser-h {
                color: #49a4a2;
            }
        }
    }

    .text-container {
        padding-top: 15px;
        padding-bottom: 15px;
        .intro {
            margin: 0;
        }
    }
}
`;

export const BreakingTeaserStyle = createGlobalStyle`
${getAllCardStyles().globalStyle.rules}
${getAllTextStyles(['breaking-teaser-h']).globalStyle.rules}
${(TeaserFigureStyle as any).globalStyle.rules}
${(TeaserRelatedStyle as any).globalStyle.rules}
${(GlobalStyle as any).globalStyle.rules}
`;
