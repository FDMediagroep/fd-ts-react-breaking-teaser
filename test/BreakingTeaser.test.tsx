import TestRenderer from 'react-test-renderer';
import BreakingTeaser from '../src/BreakingTeaser';
import React from 'react';

describe('BreakingTeaser', () => {
    test('renders correctly', () => {
        let breakingTeaser = TestRenderer.create(<BreakingTeaser
            id="123"
            bookmarked={false}
            comments={99}
            description="Containers van warenhuizen als Amazon en Alibaba bevatten een veelheid aan producten die allemaal apart ingeklaard moeten worden."
            url="https://dev.fd.nl/ondernemen/1254849/douane-20-tot-30-meer-aangiftes-door-groei-in-webwinkelen"
            hashTags="#Kitchensink"
            imageUrl="https://fd-external-development.imgix.net/43ed7a431f43ef9a3d43afd82d6ccbeb82610ffa.jpg"
            alt="Douane: 20 tot 30% meer aangiftes door groei in webwinkelen"
            onBookmark={() => {}}
            subject="Monetair"
            readableAge="7 uur"
            related={[{
                longread: true,
                title: 'Als vrouw moet je altijd onafhankelijk zijn',
                url: 'https://dev.fd.nl/profiel/1235280/als-vrouw-moet-je-onafhankelijk-zijn',
                uuid: "234"
            }, {
                title: 'President van de Amerikaanse Fed hamert op onafhankelijkheid',
                url: 'https://dev.fd.nl/economie-politiek/1235274/president-van-de-amerikaanse-fed-hamert-op-onafhankelijkheid',
                uuid: "345"
            }]}
            title="Douane: 20 tot 30% meer aangiftes door groei in webwinkelen"
        />);
        expect(breakingTeaser.toJSON()).toMatchSnapshot();
    });

    test('renders bookmarked correctly', () => {
        let breakingTeaser = TestRenderer.create(<BreakingTeaser
            id="123"
            bookmarked={true}
            comments={99}
            description="Containers van warenhuizen als Amazon en Alibaba bevatten een veelheid aan producten die allemaal apart ingeklaard moeten worden."
            url="https://dev.fd.nl/ondernemen/1254849/douane-20-tot-30-meer-aangiftes-door-groei-in-webwinkelen"
            hashTags="#Kitchensink"
            imageUrl="https://fd-external-development.imgix.net/43ed7a431f43ef9a3d43afd82d6ccbeb82610ffa.jpg"
            alt="Douane: 20 tot 30% meer aangiftes door groei in webwinkelen"
            onBookmark={() => {}}
            subject="Monetair"
            readableAge="7 uur"
            related={[{
                longread: true,
                title: 'Als vrouw moet je altijd onafhankelijk zijn',
                url: 'https://dev.fd.nl/profiel/1235280/als-vrouw-moet-je-onafhankelijk-zijn',
                uuid: "234"
            }, {
                title: 'President van de Amerikaanse Fed hamert op onafhankelijkheid',
                url: 'https://dev.fd.nl/economie-politiek/1235274/president-van-de-amerikaanse-fed-hamert-op-onafhankelijkheid',
                uuid: "345"
            }]}
            title="Douane: 20 tot 30% meer aangiftes door groei in webwinkelen"
        />);
        expect(breakingTeaser.toJSON()).toMatchSnapshot();
    });
});
