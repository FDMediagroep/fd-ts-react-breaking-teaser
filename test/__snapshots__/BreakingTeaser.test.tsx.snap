// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreakingTeaser renders bookmarked correctly 1`] = `
<div
  cardStyle="default"
  className="fd-card fd-breaking-teaser default"
>
  <div
    className="top-bar"
  >
    <span>
      Breaking
    </span>
    <span
      className="icons"
    >
      <a
        className="share-social"
        href="javascript: void(0);"
      >
        <i
          className="share-social icon-share-alt"
          data-hashtags="#Kitchensink"
          data-id="123"
          data-section="Monetair"
          data-title="Douane: 20 tot 30% meer aangiftes door groei in webwinkelen"
          data-url="https://dev.fd.nl/ondernemen/1254849/douane-20-tot-30-meer-aangiftes-door-groei-in-webwinkelen"
          title="Delen"
        />
      </a>
      <a
        className="favorite selected"
        data-addurl="/add-favorite"
        data-articleid="123"
        data-deleteurl="/delete-favorite"
        data-key="123"
        data-title="Douane: 20 tot 30% meer aangiftes door groei in webwinkelen"
        data-url="https://dev.fd.nl/ondernemen/1254849/douane-20-tot-30-meer-aangiftes-door-groei-in-webwinkelen"
        href="javascript: void(0);"
        id="favorite123"
        onClick={[Function]}
        title="Bewaren"
      >
        <i
          className="icon-bookmark"
          title="Bewaren"
        />
        <i
          className="icon-bookmark1"
          title="Verwijder van bewaarlijst"
        />
      </a>
      <a
        className="comments-link"
        href="https://dev.fd.nl/ondernemen/1254849/douane-20-tot-30-meer-aangiftes-door-groei-in-webwinkelen#comments"
      >
        <i
          className="icon-comment-alt"
        />
        <span
          className="comments-count"
        >
          99
        </span>
        <span
          className="comments-suffix"
        >
          reacties
        </span>
      </a>
    </span>
  </div>
  <a
    className="teaser-link"
    href="https://dev.fd.nl/ondernemen/1254849/douane-20-tot-30-meer-aangiftes-door-groei-in-webwinkelen"
  >
    <figure
      className="fd-teaser-figure"
    >
      <picture>
        <source
          media="(max-width: 640px)"
          srcSet="https://fd-external-development.imgix.net/43ed7a431f43ef9a3d43afd82d6ccbeb82610ffa.jpg?fit=crop&crop=faces&auto=format&fm=jpg&q=45&w=616&h=346"
        />
        <source
          media="(max-width: 860px)"
          srcSet="https://fd-external-development.imgix.net/43ed7a431f43ef9a3d43afd82d6ccbeb82610ffa.jpg?fit=crop&crop=faces&auto=format&fm=jpg&q=45&w=740&h=247"
        />
        <source
          media="(min-width: 861px)"
          srcSet="https://fd-external-development.imgix.net/43ed7a431f43ef9a3d43afd82d6ccbeb82610ffa.jpg?fit=crop&crop=faces&auto=format&fm=jpg&q=45&w=1034&h=344"
        />
        <img
          alt="Douane: 20 tot 30% meer aangiftes door groei in webwinkelen"
          src="https://fd-external-development.imgix.net/43ed7a431f43ef9a3d43afd82d6ccbeb82610ffa.jpg?fit=crop&crop=faces&auto=format&fm=jpg&q=45&w=616&h=346"
        />
      </picture>
    </figure>
    <div
      className="text-container"
    >
      <h2
        className="fd-typography breaking-teaser-h"
        textStyle="breaking-teaser-h"
      >
        Douane: 20 tot 30% meer aangiftes door groei in webwinkelen
      </h2>
      <p
        className="intro"
      >
        Containers van warenhuizen als Amazon en Alibaba bevatten een veelheid aan producten die allemaal apart ingeklaard moeten worden.
      </p>
    </div>
  </a>
  <nav
    className="related"
  >
    <a
      className="related-link longread"
      href="https://dev.fd.nl/profiel/1235280/als-vrouw-moet-je-onafhankelijk-zijn"
    >
      <span
        className="related-title"
      >
        Als vrouw moet je altijd onafhankelijk zijn
      </span>
      <i
        className="icon-chevron-right"
      />
    </a>
    <a
      className="related-link"
      href="https://dev.fd.nl/economie-politiek/1235274/president-van-de-amerikaanse-fed-hamert-op-onafhankelijkheid"
    >
      <span
        className="related-title"
      >
        President van de Amerikaanse Fed hamert op onafhankelijkheid
      </span>
      <i
        className="icon-chevron-right"
      />
    </a>
  </nav>
</div>
`;

exports[`BreakingTeaser renders correctly 1`] = `
<div
  cardStyle="default"
  className="fd-card fd-breaking-teaser default"
>
  <div
    className="top-bar"
  >
    <span>
      Breaking
    </span>
    <span
      className="icons"
    >
      <a
        className="share-social"
        href="javascript: void(0);"
      >
        <i
          className="share-social icon-share-alt"
          data-hashtags="#Kitchensink"
          data-id="123"
          data-section="Monetair"
          data-title="Douane: 20 tot 30% meer aangiftes door groei in webwinkelen"
          data-url="https://dev.fd.nl/ondernemen/1254849/douane-20-tot-30-meer-aangiftes-door-groei-in-webwinkelen"
          title="Delen"
        />
      </a>
      <a
        className="favorite"
        data-addurl="/add-favorite"
        data-articleid="123"
        data-deleteurl="/delete-favorite"
        data-key="123"
        data-title="Douane: 20 tot 30% meer aangiftes door groei in webwinkelen"
        data-url="https://dev.fd.nl/ondernemen/1254849/douane-20-tot-30-meer-aangiftes-door-groei-in-webwinkelen"
        href="javascript: void(0);"
        id="favorite123"
        onClick={[Function]}
        title="Bewaren"
      >
        <i
          className="icon-bookmark"
          title="Bewaren"
        />
        <i
          className="icon-bookmark1"
          title="Verwijder van bewaarlijst"
        />
      </a>
      <a
        className="comments-link"
        href="https://dev.fd.nl/ondernemen/1254849/douane-20-tot-30-meer-aangiftes-door-groei-in-webwinkelen#comments"
      >
        <i
          className="icon-comment-alt"
        />
        <span
          className="comments-count"
        >
          99
        </span>
        <span
          className="comments-suffix"
        >
          reacties
        </span>
      </a>
    </span>
  </div>
  <a
    className="teaser-link"
    href="https://dev.fd.nl/ondernemen/1254849/douane-20-tot-30-meer-aangiftes-door-groei-in-webwinkelen"
  >
    <figure
      className="fd-teaser-figure"
    >
      <picture>
        <source
          media="(max-width: 640px)"
          srcSet="https://fd-external-development.imgix.net/43ed7a431f43ef9a3d43afd82d6ccbeb82610ffa.jpg?fit=crop&crop=faces&auto=format&fm=jpg&q=45&w=616&h=346"
        />
        <source
          media="(max-width: 860px)"
          srcSet="https://fd-external-development.imgix.net/43ed7a431f43ef9a3d43afd82d6ccbeb82610ffa.jpg?fit=crop&crop=faces&auto=format&fm=jpg&q=45&w=740&h=247"
        />
        <source
          media="(min-width: 861px)"
          srcSet="https://fd-external-development.imgix.net/43ed7a431f43ef9a3d43afd82d6ccbeb82610ffa.jpg?fit=crop&crop=faces&auto=format&fm=jpg&q=45&w=1034&h=344"
        />
        <img
          alt="Douane: 20 tot 30% meer aangiftes door groei in webwinkelen"
          src="https://fd-external-development.imgix.net/43ed7a431f43ef9a3d43afd82d6ccbeb82610ffa.jpg?fit=crop&crop=faces&auto=format&fm=jpg&q=45&w=616&h=346"
        />
      </picture>
    </figure>
    <div
      className="text-container"
    >
      <h2
        className="fd-typography breaking-teaser-h"
        textStyle="breaking-teaser-h"
      >
        Douane: 20 tot 30% meer aangiftes door groei in webwinkelen
      </h2>
      <p
        className="intro"
      >
        Containers van warenhuizen als Amazon en Alibaba bevatten een veelheid aan producten die allemaal apart ingeklaard moeten worden.
      </p>
    </div>
  </a>
  <nav
    className="related"
  >
    <a
      className="related-link longread"
      href="https://dev.fd.nl/profiel/1235280/als-vrouw-moet-je-onafhankelijk-zijn"
    >
      <span
        className="related-title"
      >
        Als vrouw moet je altijd onafhankelijk zijn
      </span>
      <i
        className="icon-chevron-right"
      />
    </a>
    <a
      className="related-link"
      href="https://dev.fd.nl/economie-politiek/1235274/president-van-de-amerikaanse-fed-hamert-op-onafhankelijkheid"
    >
      <span
        className="related-title"
      >
        President van de Amerikaanse Fed hamert op onafhankelijkheid
      </span>
      <i
        className="icon-chevron-right"
      />
    </a>
  </nav>
</div>
`;
